src/Runner/Runner.php

```diff
@@ -27,16 +29,30 @@
 use PhpCsFixer\Linter\LinterInterface;
 use PhpCsFixer\Linter\LintingException;
 use PhpCsFixer\Linter\LintingResultInterface;
+use PhpCsFixer\Runner\Parallel\ParallelisationException;
+use PhpCsFixer\Runner\Parallel\Process;
+use PhpCsFixer\Runner\Parallel\ProcessIdentifier;
+use PhpCsFixer\Runner\Parallel\ProcessPool;
 use PhpCsFixer\Tokenizer\Tokens;
+use React\EventLoop\StreamSelectLoop;
+use React\Socket\ConnectionInterface;
+use React\Socket\TcpServer;
 use Symfony\Component\EventDispatcher\EventDispatcherInterface;
 use Symfony\Component\Filesystem\Exception\IOException;
 use Symfony\Contracts\EventDispatcher\Event;
 
 /**
  * @author Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ * @author Greg Korba <greg@codito.dev>
+ *
+ * @phpstan-type _RunResult array<string, array{appliedFixers: list<string>, diff: string}>
+ *
+ * @internal
```

Comment:
adding `@internal` is officially a BC break 🤔

Reply:
IMHO it should be internal from the beginning. Was it intended for programmatic usage? If yes, then we have bigger problem than `@internal` - modified constructor signature. I just really did not want to add another argument there, but if we don't want to change it, then it's not a big problem to revert old signature and add `ParallelConfig` as an additional, optional argument at the end. You decide 🙂.

Reply:
FYI:

- [ECS does not use `Runner`](https://github.com/easy-coding-standard/easy-coding-standard/blob/01c20b67b83b3b128cd8d77fd72b194cf8c312bd/src/FixerRunner/Application/FixerFileProcessor.php)
- [Pint uses `Runner`](https://github.com/laravel/pint/blob/fbb8dba3fe6b24f311e508792475c1e272415dd7/app/Actions/FixCode.php). @nunomaduro would it be problem for you if `Runner`'s signature was changed?

I am not aware of any other tools that use Fixer under the hood.

Reply:
indeed, changing public interface opens the doors for other issues.

> IMHO it should be internal from the beginning.

we can consider changing what's `@internal` and what's `@public` separately, and mark it with proper deprecation and change affecting next major.
It was explicitly part of contract for those using Fixer as library and not CLI tool, back in past we discussed it on GrumPHP case.
In the end, it's not if few tools we think about are relying on given part of public interface or not, but about the tools that we are not having in mind - as public is public, not "public for few repos we fingerpoint".

Reply:
@Wirone Pint uses a specific (pinpointed) version of PHPCSFixer. So, if you were to make changes to the Runner's `__construct` or anything else, we would just adapt ourselves to those changes before upgrading to the latest PHPCSFixer version. 👍🏻

Reply:
Thanks @nunomaduro for the feedback. Anyway, I think I am going to revert this signature change and will just add parallel config as new, optional argument. But we definitely should refactor this in v4.

Reply:
@Wirone , would it be possible to create a new runner (RunnerV4) class that we use everywhere internally, with clean design up to our needs, and let old Runner remain either as-is (legacy copy-paste), or be a proxy/wrapper to RunnerV4, keeping old interface?

That way, we can also add deprecation notice to Runner (the old one).

Reply:
@keradus it's possible, but I don't know if I want to do this in this PR. It's out of the scope of parallelisation to design things for v4, especially that we don't even have a vision for the next major version. I'll just keep BC and we can refactor things later. Let's just keep focus on parallel worker and provide minimal required changes for that.

Reply:
Added todo item in issue's description, so I'm going to resolve this discussion so it's easier to track what's left.